#Задание 1
#Работа с библиотекой pandas
**Цель работы:** познакомиться с использованием библиотеки pandas при решении практических задач, связанных с анализом данных.

Проанализируем датасет *home_data.csv*. Датасет содержит различную информацию о квартирах, выставленных на продажу. Набор данных включает следующие атрибуты:
- id – идентификатор;
- date – дата;
- price – цена
- bedrooms – спальни
- bathrooms – ванные комнаты
- sqft_living – жилая площадь
- sqft_lot – площадь участка
- floors – этажи
- waterfront – вид на море
- view – вид из окна
- condition – состояние
- grade – класс
- sqft_above – высота над уровнем моря
- sqft_basement – цокольный этаж
- yr_built – год постройки
- yr_renovated – год реновации
- zip code – почтовый индекс
- lat – широта
- long – долгота
- sqft_living15 – жилая площадь
- sqft_lot15 – площадь участка

##1. Импортируем необходимые библиотеки

In [6]:
import numpy as np
import pandas as pd

##2. Импортируем набор данных из файла и помещаем его в объект DataFrame

In [7]:
df = pd.read_csv("/content/home_data.csv")

##3. Выведем первые 15 строк

In [8]:
df.head(15)

id             date    price  bedrooms  bathrooms  sqft_living  \
0   7129300520  20141013T000000   221900         3       1.00         1180   
1   6414100192  20141209T000000   538000         3       2.25         2570   
2   5631500400  20150225T000000   180000         2       1.00          770   
3   2487200875  20141209T000000   604000         4       3.00         1960   
4   1954400510  20150218T000000   510000         3       2.00         1680   
5   7237550310  20140512T000000  1225000         4       4.50         5420   
6   1321400060  20140627T000000   257500         3       2.25         1715   
7   2008000270  20150115T000000   291850         3       1.50         1060   
8   2414600126  20150415T000000   229500         3       1.00         1780   
9   3793500160  20150312T000000   323000         3       2.50         1890   
10  1736800520  20150403T000000   662500         3       2.50         3560   
11  9212900260  20140527T000000   468000         2       1.00         1160   
12   114101516  20140528T000000   310000         3       1.00         1430   
13  6054650070  20141007T000000   400000         3       1.75         1370   
14  1175000570  20150312T000000   530000         5       2.00         1810   

    sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0       5650     1.0           0     0  ...      7        1180              0   
1       7242     2.0           0     0  ...      7        2170            400   
2      10000     1.0           0     0  ...      6         770              0   
3       5000     1.0           0     0  ...      7        1050            910   
4       8080     1.0           0     0  ...      8        1680              0   
5     101930     1.0           0     0  ...     11        3890           1530   
6       6819     2.0           0     0  ...      7        1715              0   
7       9711     1.0           0     0  ...      7        1060              0   
8       7470     1.0           0     0  ...      7        1050            730   
9       6560     2.0           0     0  ...      7        1890              0   
10      9796     1.0           0     0  ...      8        1860           1700   
11      6000     1.0           0     0  ...      7         860            300   
12     19901     1.5           0     0  ...      7        1430              0   
13      9680     1.0           0     0  ...      7        1370              0   
14      4850     1.5           0     0  ...      7        1810              0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0       1955             0    98178  47.5112 -122.257           1340   
1       1951          1991    98125  47.7210 -122.319           1690   
2       1933             0    98028  47.7379 -122.233           2720   
3       1965             0    98136  47.5208 -122.393           1360   
4       1987             0    98074  47.6168 -122.045           1800   
5       2001             0    98053  47.6561 -122.005           4760   
6       1995             0    98003  47.3097 -122.327           2238   
7       1963             0    98198  47.4095 -122.315           1650   
8       1960             0    98146  47.5123 -122.337           1780   
9       2003             0    98038  47.3684 -122.031           2390   
10      1965             0    98007  47.6007 -122.145           2210   
11      1942             0    98115  47.6900 -122.292           1330   
12      1927             0    98028  47.7558 -122.229           1780   
13      1977             0    98074  47.6127 -122.045           1370   
14      1900             0    98107  47.6700 -122.394           1360   

    sqft_lot15  
0         5650  
1         7639  
2         8062  
3         5000  
4         7503  
5       101930  
6         6819  
7         9711  
8         8113  
9         7570  
10        8925  
11        6000  
12       12697  
13       10208  
14        4850  

[15 rows x 21 columns]

##4. Сколько строк и столбцов в наборе данных?


In [10]:
rows = df.shape[0]
columns = df.shape[1]
rows, columns

(21613, 21)

##5. Выведем имена всех столбцов и их типы

In [11]:
df.dtypes

id                 int64
date              object
price              int64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

##6. Получим сводную информацию по датафрейму

In [12]:
df.info(0)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  int64  
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

##7. Что является индексом этого набора данных? Проиндексируйте набор данных по полю id

In [13]:
df.index

RangeIndex(start=0, stop=21613, step=1)

In [14]:
copy = df.set_index("id")
copy.index

Index([7129300520, 6414100192, 5631500400, 2487200875, 1954400510, 7237550310,
       1321400060, 2008000270, 2414600126, 3793500160,
       ...
       7852140040, 9834201367, 3448900210, 7936000429, 2997800021,  263000018,
       6600060120, 1523300141,  291310100, 1523300157],
      dtype='int64', name='id', length=21613)

##8. Преобразуем значения price в числа с плавающей запятой

In [15]:
df["price"] = df['price'].astype("float64")

In [16]:
df.info(0)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

##9. Вычислим для столбца price среднее значение, моду и медиану

In [36]:
mean_price = df['price'].mean()   #среднее значение
mode_price = df['price'].mode()[0]   #мода
median_price = df['price'].median(0) #медиана
print(f'Среднее значение: {mean_price}, мода: {mode_price}, медиана: {median_price}')

Среднее значение: 540088.1419053348, мода: 350000.0, медиана: 450000.0


##10. Определим для столбца price дисперсию и среднеквадратичное отклонение

In [37]:
variance_price = df['price'].var()
std_dev_price = df['price'].std()
print(f'Дисперсия: {variance_price}, среднеквадратичное отклонение: {std_dev_price}')

Дисперсия: 134782378019.59238, среднеквадратичное отклонение: 367127.1959683624


##11. Как влияет высота над уровнем моря на цену жилья?

In [56]:
corr = df['sqft_above'].corr(df['price'])
corr

np.float64(0.6055672952130322)

Воспользуемся встроенным расчётом корреляции Пирсона.

Значение 0.606 означает среднюю связь, оно положительно, значит связь прямая. Есть зависимость - чем больше высота над уровнем моря, тем больше цена.

##12. Определим TOP-5 самых дорогих и TOP-5 самых дешевых квартир

Топ-5 самых дорогих квартир

In [50]:
df_copy = df.copy()
df_copycopy = df_copy.sort_values(by='price', ascending=False)
df_copycopy.head(5)

id             date      price  bedrooms  bathrooms  \
7252  6762700020  20141013T000000  7700000.0         6       8.00   
3914  9808700762  20140611T000000  7062500.0         5       4.50   
9254  9208900037  20140919T000000  6885000.0         6       7.75   
4411  2470100110  20140804T000000  5570000.0         5       5.75   
1448  8907500070  20150413T000000  5350000.0         5       5.00   

      sqft_living  sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
7252        12050     27600     2.5           0     3  ...     13        8570   
3914        10040     37325     2.0           1     2  ...     11        7680   
9254         9890     31374     2.0           0     4  ...     13        8860   
4411         9200     35069     2.0           0     0  ...     13        6200   
1448         8000     23985     2.0           0     4  ...     12        6720   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
7252           3480      1910          1987    98102  47.6298 -122.323   
3914           2360      1940          2001    98004  47.6500 -122.214   
9254           1030      2001             0    98039  47.6305 -122.240   
4411           3000      2001             0    98039  47.6289 -122.233   
1448           1280      2009             0    98004  47.6232 -122.220   

      sqft_living15  sqft_lot15  
7252           3940        8800  
3914           3930       25449  
9254           4540       42730  
4411           3560       24345  
1448           4600       21750  

[5 rows x 21 columns]

Топ-5 самых дешёвых квартир

In [51]:
df_copy = df.copy()
df_copycopy = df_copy.sort_values(by='price', ascending=True)
df_copycopy.head(5)

id             date    price  bedrooms  bathrooms  sqft_living  \
1149   3421079032  20150217T000000  75000.0         1       0.00          670   
15293    40000362  20140506T000000  78000.0         2       1.00          780   
465    8658300340  20140523T000000  80000.0         1       0.75          430   
16198  3028200080  20150324T000000  81000.0         2       1.00          730   
8274   3883800011  20141105T000000  82000.0         3       1.00          860   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
1149      43377     1.0           0     0  ...      3         670   
15293     16344     1.0           0     0  ...      5         780   
465        5050     1.0           0     0  ...      4         430   
16198      9975     1.0           0     0  ...      5         730   
8274      10426     1.0           0     0  ...      6         860   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
1149               0      1966             0    98022  47.2638 -121.906   
15293              0      1942             0    98168  47.4739 -122.280   
465                0      1912             0    98014  47.6499 -121.909   
16198              0      1943             0    98168  47.4808 -122.315   
8274               0      1954             0    98146  47.4987 -122.341   

       sqft_living15  sqft_lot15  
1149            1160       42882  
15293           1700       10387  
465             1200        7500  
16198            860        9000  
8274            1140       11250  

[5 rows x 21 columns]

##13. Получим данные по квартирам, подвергавшимся реновации


In [47]:
df_renovated = df.copy
df_renovated = df.drop(df[df['yr_renovated'] == 0].index)

In [49]:
df_renovated

id             date      price  bedrooms  bathrooms  \
1      6414100192  20141209T000000   538000.0         3       2.25   
35     9547205180  20140613T000000   696000.0         3       2.50   
95     1483300570  20140908T000000   905000.0         4       2.50   
103    2450000295  20141007T000000  1088000.0         3       2.50   
115    3626039325  20141121T000000   740500.0         3       3.50   
...           ...              ...        ...       ...        ...   
19622  7351200295  20150114T000000  1150000.0         3       1.75   
20057   126039256  20140904T000000   434900.0         3       2.00   
20444  4305600360  20150225T000000   500012.0         4       2.50   
20447  3319500628  20150212T000000   356999.0         3       1.50   
20962  1278000210  20150311T000000   110000.0         2       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
1             2570      7242     2.0           0     0  ...      7   
35            2300      3060     1.5           0     0  ...      8   
95            3300     10250     1.0           0     0  ...      7   
103           2920      8113     2.0           0     0  ...      8   
115           4380      6350     2.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
19622         1760      6788     2.0           1     4  ...      7   
20057         1520      5040     2.0           0     0  ...      7   
20444         2400      9612     1.0           0     0  ...      8   
20447         1010      1546     2.0           0     0  ...      8   
20962          828      4524     1.0           0     0  ...      6   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
1            2170            400      1951          1991    98125  47.7210   
35           1510            790      1930          2002    98115  47.6827   
95           2390            910      1946          1991    98040  47.5873   
103          2920              0      1950          2010    98004  47.5814   
115          2780           1600      1900          1999    98117  47.6981   
...           ...            ...       ...           ...      ...      ...   
19622        1760              0      1940          1960    98125  47.7336   
20057        1520              0      1977          2006    98177  47.7770   
20444        1230           1170      1962          2009    98059  47.4799   
20447        1010              0      1971          2014    98144  47.5998   
20962         828              0      1968          2007    98001  47.2655   

          long  sqft_living15  sqft_lot15  
1     -122.319           1690        7639  
35    -122.310           1590        3264  
95    -122.249           1950        6045  
103   -122.196           2370        8113  
115   -122.368           1830        6350  
...        ...            ...         ...  
19622 -122.284           1630        7588  
20057 -122.362           1860        8710  
20444 -122.127           2430        5539  
20447 -122.311           1010        1517  
20962 -122.244            828        5402  

[914 rows x 21 columns]

##14. Найдём жильё самого старого и самого нового года постройки

Жильё самого старого года постройки

In [52]:
df_old = df.copy()
df_oldest = df_copy.sort_values(by='yr_built', ascending=True)
df_oldest.head(1)

id             date     price  bedrooms  bathrooms  sqft_living  \
3882  7129302095  20150213T000000  265000.0         3        1.0         1122   

      sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
3882      6554     1.5           0     0  ...      5        1122   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
3882              0      1900             0    98118  47.5135 -122.257   

      sqft_living15  sqft_lot15  
3882           1610        5650  

[1 rows x 21 columns]

Жильё самого нового года постройки

In [55]:
df_old = df.copy()
df_oldest = df_copy.sort_values(by='yr_built', ascending=False)
df_oldest.head(1)

id             date     price  bedrooms  bathrooms  sqft_living  \
20698  255470030  20150429T000000  619990.0         4       2.75         2630   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
20698      4501     2.0           0     0  ...      8        2630   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
20698              0      2015             0    98028  47.7748 -122.244   

       sqft_living15  sqft_lot15  
20698           2380        4599  

[1 rows x 21 columns]

##15.Как влияет площадь участка на стоимость жилья?

In [60]:
corr = df['sqft_lot15'].corr(df['price'])
corr

np.float64(0.0824471517581486)

Воспользуемся встроенным расчётом корреляции Пирсона.

Значение **0.082** означает слабую связь или её отсутствие. Площадь участка почти никак не влияет на стоимость жилья.